File: solutions/TypeScript/[0200_Medium] Number of Islands/numIslands.ts
function numIslands(grid: string[][]): number {
  const height = grid.length;
  const width = grid[0].length;

  const isValidPosition = (
    y: number,
    x: number
  ): boolean => (
    y >= 0 && y < height && x >= 0 && x < width
  );

  const visitIsland = (
    y: number,
    x: number
  ): void => {
    if (!isValidPosition(y, x) || grid[y][x] === '0') {
      return;
    }

    grid[y][x] = '0';

    visitIsland(y + 1, x);
    visitIsland(y - 1, x);
    visitIsland(y, x + 1);
    visitIsland(y, x - 1);
  };

  let result = 0;

  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      if (grid[y][x] === '1') {
        result++;
        visitIsland(y, x);
      }
    }
  }

  return result;
}

File: solutions/Java/[0200_Medium] Number of Islands/numIslands.java
class Solution {
  private boolean isValidPosition(int y, int x, int height, int width) {
    return y >= 0 && y < height && x >= 0 && x < width;
  }

  private void visitIsland(char[][] grid, int y, int x, int height, int width) {
    if (!isValidPosition(y, x, height, width) || grid[y][x] == '0') {
      return;
    }

    grid[y][x] = '0';

    visitIsland(grid, y + 1, x, height, width);
    visitIsland(grid, y - 1, x, height, width);
    visitIsland(grid, y, x + 1, height, width);
    visitIsland(grid, y, x - 1, height, width);
  }

  public int numIslands(char[][] grid) {
    int height = grid.length;
    int width = grid[0].length;

    int result = 0;

    for (int y = 0; y < height; y++) {
      for (int x = 0; x < width; x++) {
        if (grid[y][x] == '1') {
          result++;
          visitIsland(grid, y, x, height, width);
        }
      }
    }

    return result;
  }
}
